src/Sema.zig

```diff
@@ -17499,29 +17502,269 @@ fn resolvePeerTypes(
                 },
                 else => {},
             },
-            .ErrorUnion => {
-                const payload_ty = candidate_ty.errorUnionPayload();
-                if (chosen_ty_tag == .Pointer and
-                    chosen_ty.ptrSize() == .One and
-                    chosen_ty.childType().zigTypeTag() == .Array and
-                    payload_ty.isSlice())
-                {
-                    const chosen_child_ty = chosen_ty.childType();
-                    const chosen_elem_ty = chosen_child_ty.elemType2();
-                    const candidate_elem_ty = payload_ty.elemType2();
-                    if ((try sema.coerceInMemoryAllowed(block, candidate_elem_ty, chosen_elem_ty, false, target, src, src)) == .ok) {
+            .ErrorSet => {
+                if (chosen_ty_tag == .ErrorSet) {
+                    assert(err_set_ty != null);
+
+                    // If chosen type is anyerror, then we can use the prev type
+                    if (err_set_ty.?.isAnyError()) continue;
+
+                    // At this point, we must resolve any inferred error sets
+                    if (candidate_ty.castTag(.error_set_inferred)) |inferred| {
+                        try sema.resolveInferredErrorSet(inferred.data);
+                    }
+
+                    // If candidate is anyerror then we use it because it
+                    // is trivially a supserset of previous error set
+                    if (candidate_ty.isAnyError()) {
+                        err_set_ty = candidate_ty;
                         chosen = candidate;
                         chosen_i = candidate_i + 1;
+                        continue;
+                    }
 
-                        convert_to_slice = false; // it already is a slice
+                    // If chosen is superset of candidate, keep it.
+                    // If candidate is superset of chosen, switch it.
+                    // If neither is a superset, merge errors.
+                    var prev_is_superset = true;
+                    for (candidate_ty.errorSetNames()) |name| {
+                        if (!err_set_ty.?.errorSetHasField(name)) {
+                            prev_is_superset = false;
+                            break;
+                        }
+                    }
+                    if (prev_is_superset) continue; // use previous
```

Comment:
Instead of a pattern like this you should label the outer loop and use the labeled continue syntax.

Reply:
Sorry, can you clarify what you're asking? Would you prefer this?

```zig
var prev_is_superset = blk: {
    for (candidate_ty.errorSetNames()) |name| 
        if (!err_set_ty.?.errorSetHasField(name)) break :blk false;

    break :blk true;
}
if (prev_is_superset) continue; 
```

Reply:
I meant:
```zig
candidate_loop: for (instructions[1..]) |candidate, candidate_i| {
...
for (candidate_ty.errorSetNames()) |name| {
    if (!err_set_ty.?.errorSetHasField(name)) {
        continue :main_loop; // use previous
    }
}
```
Using a flag variable like this is not necessary in Zig thanks to labeled loops allowing breaking from nested loops.

Reply:
Oh yes, I would do that, but the logic is opposite: we only want to continue the main loop if we make it all the way through (_not_ break).

Reply:
Actually, I think I've identified a way to clean this up.

Reply:
Oh I see, in that case you'll want to use the else branch:
```zig
for (candidate_ty.errorSetNames()) |name| {
    if (!err_set_ty.?.errorSetHasField(name)) break;
} else continue :candidate_loop;
```

Reply:
Ah hah! I misunderstood `else` to only work for optionals, not generally for any loop done semantics. Perfect. I updated it, tests pass, and pushed the commit.